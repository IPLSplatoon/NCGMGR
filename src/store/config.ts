import { createStore, Store, useStore } from 'vuex'
import { InjectionKey } from 'vue'

export const LOCAL_STORAGE_CONFIG_KEY = 'config'

export interface Configuration {
    installPath: string
}

export const configStore = createStore<Configuration>({
    state: {
        installPath: ''
    },
    mutations: {
        setInstallPath: (state, installPath: string) => {
            state.installPath = installPath
        },
        setState: (state, config: Configuration) => {
            Object.assign(state, config)
        }
    },
    actions: {
        load: store => {
            const config = JSON.parse(localStorage.getItem(LOCAL_STORAGE_CONFIG_KEY) ?? '{}')
            store.commit('setState', {
                ...store.state,
                ...config
            })
        },
        persist: store => {
            localStorage.setItem(LOCAL_STORAGE_CONFIG_KEY, JSON.stringify(store.state))
        }
    }
})

export const configStoreKey: InjectionKey<Store<Configuration>> = Symbol()

export function useConfigStore (): Store<Configuration> {
    return useStore(configStoreKey)
}
