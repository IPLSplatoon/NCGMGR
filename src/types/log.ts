export interface LogEvent {
    message: string
    type?: 'success' | 'info' | 'error'
}

export interface ProgressEvent {
    step: number
    // eslint-disable-next-line camelcase
    max_step: number
}

export enum ActionState {
    COMPLETED_SUCCESS,
    COMPLETED_ERROR,
    INCOMPLETE
}
