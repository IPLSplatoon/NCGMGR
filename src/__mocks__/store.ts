import { createStore, Store } from 'vuex'
import { Configuration } from '@/store/config'
import { LogStore } from '@/store/log'

export function createConfigStore (): Store<Configuration> {
    return createStore<Configuration>({
        state: {
            installPath: '/install/path'
        },
        actions: {
            load: jest.fn(),
            persist: jest.fn()
        },
        mutations: {
            setInstallPath: jest.fn()
        }
    })
}

export function createLogStore (): Store<LogStore> {
    return createStore<LogStore>({
        state: {
            lines: [],
            completed: false
        },
        mutations: {
            reset: jest.fn(),
            insertLine: jest.fn(),
            setCompleted: jest.fn()
        }
    })
}
