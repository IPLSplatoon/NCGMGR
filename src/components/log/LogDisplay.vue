<template>
    <div class="log-display">
        <span
            v-for="(line, i) in log"
            :key="`log-line_${i}`"
            :class="`type-${line.type ?? 'unknown'}`"
        >
            <template
                v-for="(part, j) in line.messageParts"
            >
                <template v-if="!part.was_processed">
                    {{ part.content }}
                </template>
                <span
                    v-else
                    :key="`log-line_${i}_${j}`"
                    :class="classListFromLogLine(part)"
                >
                    {{ part.content }}
                </span>
            </template>
        </span>
    </div>
</template>

<script lang="ts">
import { computed, defineComponent, onMounted, onUnmounted, watch } from 'vue'
import { useLogStore } from '@/store/logStore'
import Anser from 'anser'

export default defineComponent({
    name: 'LogDisplay',

    props: {
        logKey: {
            type: String,
            required: true
        }
    },

    setup (props) {
        const logStore = useLogStore()

        watch(() => props.logKey, (newValue, oldValue) => {
            if (oldValue) {
                logStore.unlisten(oldValue)
            }
            logStore.listen(newValue)
        })

        onMounted(() => {
            logStore.listen(props.logKey)
        })

        onUnmounted(() => {
            logStore.unlisten(props.logKey)
        })

        return {
            log: computed(() => {
                const log = logStore.lines[props.logKey] ?? []
                return log.map(line => ({
                    ...line,
                    messageParts: Anser.ansiToJson(line.message, { use_classes: true })
                }))
            }),
            classListFromLogLine(line: Anser.AnserJsonEntry) {
                const result = []

                if (line.bg != null) {
                    result.push(`${line.bg}-bg`)
                }
                if (line.fg != null) {
                    result.push(`${line.fg}-fg`)
                }

                return result
            }
        }
    }
})
</script>

<style lang="scss" scoped>
@import '../../styles/layout';
@import '../../styles/text';
@import '../../styles/colors';

.log-display {
    font-family: monospace;
    width: calc(100% - 8px);
    max-height: 350px;
    text-align: left;
    border: 1px solid $input-color;
    padding: 4px 2px;
    min-height: 64px;
    overflow: scroll;
    color: var(--ipl-body-text-color);
    background-color: var(--ipl-bg-primary);

    > span {
        margin-left: 6px;
        display: block;
        white-space: nowrap;

        &.type-error {
            color: $error-color;
        }

        .foo {
            color: red;
        }
    }
}
</style>
