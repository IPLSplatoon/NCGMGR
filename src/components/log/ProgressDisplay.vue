<template>
    <ipl-progress-bar
        v-if="progress != null"
        :value="progress"
        :color="progressBarColor"
        background-color="primary"
    />
</template>

<script lang="ts">
import { defineComponent } from 'vue'
import { useLogStore } from '@/store/logStore'
import { computed } from 'vue'
import { IplProgressBar } from '@iplsplatoon/vue-components'
import { ActionState } from '@/types/log'
import { themeColors } from '@/styles/colors'

export default defineComponent({
    name: 'ProgressDisplay',

    components: { IplProgressBar },

    props: {
        logKey: {
            type: String,
            required: true
        }
    },

    setup (props) {
        const logStore = useLogStore()
        const actionState = computed(() => logStore.actionStates[props.logKey])

        return {
            progress: computed(() => {
                const entry = logStore.progressEntries[props.logKey]
                if (entry?.maxStep == null) {
                    return null
                }

                if (actionState.value != null && actionState.value !== ActionState.INCOMPLETE) {
                    return 100
                }

                return !entry ? 0 : ((entry.step ?? 0) / entry.maxStep) * 100
            }),
            progressBarColor: computed(() => {
                switch (actionState.value) {
                    case ActionState.COMPLETED_ERROR:
                        return themeColors.red
                    case ActionState.COMPLETED_SUCCESS:
                        return themeColors.green
                    default:
                        return themeColors.blue
                }
            })
        }
    }
})
</script>
