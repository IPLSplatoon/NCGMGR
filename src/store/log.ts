import { LogEvent } from '@/types/log'
import { createStore, Store, useStore } from 'vuex'
import { InjectionKey } from 'vue'
import { Event, listen, UnlistenFn } from '@tauri-apps/api/event'

export interface LogStore {
    lines: Record<string, LogEvent[]>
    completed: Record<string, boolean>
}

const unlistenFns: Record<string, UnlistenFn> = {}

export const logStore = createStore<LogStore>({
    state: {
        lines: {},
        completed: {}
    },
    mutations: {
        insertLine (state, { line, key }: { line: LogEvent, key: string }) {
            if (!state.lines[key]) {
                state.lines[key] = []
            }
            state.lines[key].push(line)
        },
        reset (state, key: string) {
            state.lines[key] = []
            state.completed[key] = false
        },
        setCompleted (state, { key, completed }: { key: string, completed: boolean }) {
            state.completed[key] = completed
        }
    },
    actions: {
        listen (store, key: string) {
            if (unlistenFns[key]) {
                return
            }
            return listen(`log:${key}`, (event: Event<LogEvent>) => {
                store.commit('insertLine', { line: event.payload, key })
            }).then(unlisten => {
                unlistenFns[key] = unlisten
            })
        },
        unlisten (store, key: string) {
            if (unlistenFns[key]) {
                unlistenFns[key]()
            }
        },
        logPromiseResult (store, { promise, key, noLogOnSuccess }: { promise: Promise<unknown>, key: string, noLogOnSuccess: boolean }) {
            promise.then(() => {
                if (!noLogOnSuccess) {
                    store.commit('insertLine', {
                        line: {
                            message: 'Success!',
                            type: 'success'
                        },
                        key
                    })
                }
            }).catch(e => {
                store.commit('insertLine', {
                    line: {
                        message: String(e),
                        type: 'error'
                    },
                    key
                })
            }).finally(() => {
                store.commit('setCompleted', { key, completed: true })
            })
        }
    }
})

export const logStoreKey: InjectionKey<Store<LogStore>> = Symbol()

export function useLogStore (): Store<LogStore> {
    return useStore(logStoreKey)
}
