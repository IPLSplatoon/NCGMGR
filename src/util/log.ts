import { logStore } from '@/store/log'

export function logPromiseResult (promise: Promise<unknown>): void {
    promise.then(() => {
        logStore.commit('insertLine', {
            message: 'Success!',
            type: 'success'
        })
    }).catch(e => {
        logStore.commit('insertLine', {
            message: String(e),
            type: 'error'
        })
    }).finally(() => {
        logStore.commit('setCompleted', true)
    })
}
