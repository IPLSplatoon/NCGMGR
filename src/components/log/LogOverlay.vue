<template>
    <mgr-overlay
        v-model:visible="modelVisible"
        class="log-overlay"
        :no-background-close="!completed"
        :max-width="moreDetailsShown"
    >
        <div>
            <ipl-dialog-title
                :title="title"
                class="m-b-8"
                :closing-disabled="!completed"
                @close="modelVisible = false"
            />
            <ipl-space>
                <div class="m-b-8">
                    {{ progressMessage }}
                </div>
                <progress-display :log-key="logKey" />
            </ipl-space>
            <ipl-expanding-space
                v-model:expanded="moreDetailsShown"
                title="More details"
                class="m-t-8 max-width"
            >
                <log-display
                    :log-key="logKey"
                    class="m-t-8"
                />
            </ipl-expanding-space>
        </div>
    </mgr-overlay>
</template>

<script lang="ts">
import { computed, defineComponent, ref } from 'vue'
import MgrOverlay from '@/components/mgr/MgrOverlay.vue'
import { IplDialogTitle, IplExpandingSpace, IplSpace } from '@iplsplatoon/vue-components'
import { useLogStore } from '@/store/logStore'
import LogDisplay from '@/components/log/LogDisplay.vue'
import ProgressDisplay from '@/components/log/ProgressDisplay.vue'
import { ActionState } from '@/types/log'

export default defineComponent({
    name: 'LogOverlay',

    components: { IplSpace, IplDialogTitle, IplExpandingSpace, ProgressDisplay, LogDisplay, MgrOverlay },

    props: {
        title: {
            type: String,
            required: true
        },
        visible: {
            type: Boolean,
            required: true
        },
        logKey: {
            type: String,
            required: true
        }
    },

    emits: ['update:visible'],

    setup (props, { emit }) {
        const logStore = useLogStore()

        const modelVisible = computed({
            get () {
                return props.visible
            },
            set (newValue: boolean) {
                emit('update:visible', newValue)
            }
        })
        const completed = computed(() => logStore.actionStates[props.logKey] !== ActionState.INCOMPLETE)

        return {
            progressMessage: computed(() => {
                switch (logStore.actionStates[props.logKey]) {
                    case ActionState.COMPLETED_ERROR:
                        return 'An error has occurred.'
                    case ActionState.COMPLETED_SUCCESS:
                        return 'Done!'
                    default:
                        return 'Please wait...'
                }
            }),
            moreDetailsShown: ref(false),
            completed,
            modelVisible
        }
    }
})
</script>
