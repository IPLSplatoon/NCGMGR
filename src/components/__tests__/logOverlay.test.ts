import LogOverlay from '@/components/logOverlay.vue'
import { useLogStore } from '@/store/log'
import { config, mount } from '@vue/test-utils'
import { IplButton } from '@iplsplatoon/vue-components'
import { createTestingPinia, TestingPinia } from '@pinia/testing'

describe('LogOverlay', () => {
    let pinia: TestingPinia

    beforeEach(() => {
        pinia = createTestingPinia()
        config.global.plugins = [pinia]
        const logStore = useLogStore()
        logStore.listen = jest.fn()
        logStore.unlisten = jest.fn()
    })

    config.global.stubs = {
        IplButton: true
    }

    it('matches snapshot', () => {
        const wrapper = mount(LogOverlay, {
            props: {
                visible: true,
                title: 'Log Overlay',
                logKey: 'log1'
            }
        })

        expect(wrapper.html()).toMatchSnapshot()
    })

    it('disables close button if log state is not completed', () => {
        const store = useLogStore()
        store.completed.logKey = false
        const wrapper = mount(LogOverlay, {
            props: {
                visible: true,
                title: 'Log Overlay',
                logKey: 'logKey'
            }
        })

        expect(wrapper.getComponent('[data-test="close-button"]').attributes().disabled).toEqual('true')
    })

    it('enables close button if log state is completed', () => {
        const store = useLogStore()
        store.completed['log-key'] = true
        const wrapper = mount(LogOverlay, {
            props: {
                visible: true,
                title: 'Log Overlay',
                logKey: 'log-key'
            }
        })

        expect(wrapper.getComponent('[data-test="close-button"]').attributes().disabled).toEqual('false')
    })

    it('closes overlay on close button click', () => {
        const store = useLogStore()
        store.completed['key-1'] = true
        const wrapper = mount(LogOverlay, {
            props: {
                visible: true,
                title: 'Log Overlay',
                logKey: 'key-1'
            }
        })

        wrapper.getComponent<typeof IplButton>('[data-test="close-button"]').vm.$emit('click')

        const emitted = wrapper.emitted('update:visible')
        expect(emitted?.length).toEqual(1)
        expect(emitted?.[0]).toEqual([false])
    })
})
