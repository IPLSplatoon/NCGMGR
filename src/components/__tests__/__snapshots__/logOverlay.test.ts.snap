// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LogOverlay matches snapshot 1`] = `
<div class="ipl-overlay__wrapper layout horizontal center-horizontal center-vertical log-overlay">
  <div class="ipl-overlay__content">
    <div class="layout vertical center-horizontal">
      <h2 class="m-b-8">Log Overlay</h2>
      <div class="log-display m-b-8"></div>
      <ipl-button-stub label="Close" color="blue" disabled="true" async="false" progressmessage="Working..." successmessage="Done!" requiresconfirmation="false" shortconfirmationmessage="false" small="false" disableonsuccess="false" href="" data-test="close-button"></ipl-button-stub>
    </div>
  </div>
</div>
`;
